### ISSUE TITLE
[ruff+ty]: No error on invalid syntax in individual cells in Jupyter notebooks

### ISSUE THREAD

**Issue by hamdanal** [astral-sh/ruff#23782]
### Summary

Ruff (and ty) does not verify that the syntax of individual cells in a Jupyter notebook is valid. It seems to treat the entire notebook as a single source.

<img width="666" height="526" alt="Image" src="https://github.com/user-attachments/assets/32d5f148-5a5e-47d1-94fc-3489fedb1ae0" />

### Version

ruff 0.15.5 and ty 0.0.21

**Comment by MichaReiser** [comment]
Thank you. 

Yes, Ruff and ty internally concatenate the source text, which can lead to the false negatives you're seeing here. 

I'm not 100% sure what the solution is; we might need to make the lexer cell-boundary aware

**Comment by ai-mountain** [comment]
@MichaReiser  could you consider my suggestion?

### Approach: Per-cell validation as a post-processing step

Rather than invasively modifying the lexer/parser, we **parse each cell individually after the concatenated parse** to catch syntax errors that the concatenation masks. The concatenated parse is still kept for cross-cell analysis (imports, variables, etc.).

### Steps

**1. Add per-cell validation in `ruff_linter` (for ruff)**
- In `crates/ruff_linter/src/linter.rs`, after the main concatenated parse, if the source is a notebook:
  - Use `CellOffsets` to extract each cell's text from the concatenated source
  - Parse each cell individually with `parse_unchecked(cell_text, options)`
  - If a cell has parse errors that the concatenated parse missed, emit them as diagnostics
  - Adjust error offsets by adding the cell's start offset

**2. Add per-cell validation in `ruff_db` (for ty)**
- In `crates/ruff_db/src/parsed.rs`, in `parsed_module_impl`, do the same per-cell validation
- Append per-cell parse errors to the `Parsed` struct (may need a `with_additional_errors` method on `Parsed`)

**3. Deduplication**
- Only report per-cell errors for cells where the concatenated parse did **not** already report an error (avoid double-reporting)

**4. Edge cases handled correctly by design**
- Multi-line strings spanning cells → flagged as unterminated string (correct, Jupyter would reject too)
- `if True:` in one cell, body in next → flagged as missing body (correct)
- Empty/magic cells → already filtered out by `is_valid_python_code_cell()`

**5. Tests**
- Notebook with `x = (1 +` in cell 1, `2)` in cell 2 → error in both cells
- Notebook with unterminated triple-quote string across cells → error
- Valid notebook → no false positives
- Single-cell notebook with syntax error → still detected (regression)
- Concatenated parse already fails → no duplicate errors

### Key files to modify
| File | Change |
|------|--------|
| `crates/ruff_linter/src/linter.rs` | Add per-cell validation after concatenated parse |
| `crates/ruff_notebook/src/notebook.rs` | Possibly add helper to extract individual cell text |
| `crates/ruff_db/src/parsed.rs` | Per-cell validation for ty |
| `crates/ruff_python_parser/src/lib.rs` | Possibly add `Parsed::with_additional_errors` |
| Test fixtures | Add `.ipynb` test files |

---

Can I proceed in this way and make a PR?

**Comment by MichaReiser** [comment]
@RenzoMXD did you use an LLM to write this summary? If so, can you please summarize and explain the change using your own words (see our [AI policy](https://github.com/astral-sh/.github/blob/main/AI_POLICY.md))?

**Comment by ai-mountain** [comment]
The problem is that Ruff and ty concatenate all notebook cells into one whole text and treat it as a single text when checking. But actually, a Jupiter Notebook consists of individual cells and executes code cell by cell, so we need validation on individual cells.

My suggestion is to divide the concatenated text into cells and perform the check on each cell individually. 
For this, in Ruff we use `CellOffsets` to get each cell's text and use the `parse_unchecked` function to check each one independently. After that, we send any syntax errors per cell. 
In ty, we perform the same method in `crates/ruff_db/src/parsed.rs`, where the notebook source is also parsed, adding per-cell validation after the concatenated parse.

**Comment by MichaReiser** [comment]
Thanks @RenzoMXD for summarising. I don't think that cells can be checked separately. For example, the following is expected to work:

Cell 1:

```py
a = 10
```

Cell 2:

```py
print(a)
```

However, ty and Ruff will warn about `a` being undefined if the cells are checked separately.

Unless you're suggesting to parse each cell separately, but still run everything else on the concatenated notebook. The issue with that is that the parsed cell AST and the parsed concatenated AST might disagree, which could lead to weird inconsistencies or duplicated errors.

**Comment by ai-mountain** [comment]
Thanks for the feedback, I agree that linting and type-checking must use the concatenated text to keep the context.

One approach I was considering: what if we parse each cell separately **only for syntax error detection**, and then send only the errors in each cell parsing but not from the concatenated parse? That way the concatenated AST is still the single text for everything else (linting, type-checking, diagnostics)

I think this solution does not solve every aspects, but it makes improvement.

**Comment by MichaReiser** [comment]
As said above. This will lead to inconsistencies between how Ruff/ty understands the code during linting and type checking and how it is understood during parsing. This can lead to conflicting diagnostics, which is something we should avoid.

**Comment by ai-mountain** [comment]
I understand the concern about conflicting diagnostics. What approach would you consider acceptable for this issue? Or do you think this is something that can't be addressed without a larger architectural change to how notebooks are handled?

**Comment by MichaReiser** [comment]
I don't have a concrete proposal. I outlined a few ideas above, but haven't had the time to explore them or consider alternatives. I consider this part of solving this issue

**Comment by chinar-amrutkar** [comment]
I think the right approach is making the lexer cell-boundary aware rather than re-parsing each cell. The lexer already tracks bracket depth, string termination state, and indentation context. At each cell boundary, we can check:

1. Unmatched brackets - If bracket depth > 0 at a cell boundary, the cell has unclosed brackets. Report at the cell's boundary position.
2. Unterminated strings - If the lexer is inside a string literal at a cell boundary, it's unterminated.
3. Incomplete compound statements - This is trickier since it requires parser-level awareness, but an if True: at the end of a cell with no body in the same cell is a real syntax error in Jupyter.

This approach avoids conflicting diagnostics because:
• It works within the existing concatenated parse (no second AST)
• It only catches errors that are genuinely masked by concatenation
• The concatenated AST remains the single source of truth for linting/type-checking

The tradeoff is that not every cell-boundary syntax error is detectable at the lexer level (e.g., x = 1 + at a boundary). But the most common and visible cases (e.g. unmatched brackets, unterminated strings) are handled without problems.

For the parser level cases (incomplete expressions), one option is to parse each cell independently but only report errors in regions where the concatenated parse didn't already report errors. This avoids duplicate diagnostics while catching more cases.

@MichaReiser - what do you think?

**Comment by MichaReiser** [comment]
Overall that's where I'm leaning towards too, but I still think there's something we need between parser and lexer to avoid cross-cell parsing.

**Comment by chinar-amrutkar** [comment]
That makes sense. The way I'm thinking about it is that the layer between lexer and parser is the token stream. The lexer already knows its state (bracket depth, string context, indent level). At a cell boundary, if the lexer is in a "dirty" state - unmatched brackets, unterminated string, mid-expression - it can inject a synthetic token (e.g., an error token or a forced newline in a context where newlines aren't allowed) that causes the parser to naturally fail at that point.

For example:

• x = (1 + at end of cell 1 - lexer is at bracket depth 1 when it hits the boundary. Injecting a synthetic newline at depth > 0 outside of parens would work, but parens allow newlines. Instead, injecting a synthetic ERRORTOKEN at the boundary would cause the parser to report a syntax error at the cell boundary position.
• """multi at end of cell 1 - lexer is in a string literal at the boundary. It can emit a synthetic unterminated string error directly, since the string state is purely a lexer concern.

This avoids both problems: 
1. no second AST (no conflicting diagnostics)
2. no cross-cell parsing (the parser sees the error token and fails). 

The concatenated parse produces the error at the right position.

This however doesn't catch cases of incomplete expressions like x = 1 + at a boundary where brackets are balanced. For those, the parser might still create a cross-cell AST node. But these are rarer than unmatched brackets and unterminated strings, and catching them might require a more detailed change (e.g., a post-parse AST walk checking node spans against cell boundaries).

Would it make sense for me to start with the lexer-level checks (brackets, strings) and iterate from there?

**Comment by MichaReiser** [comment]
I don't think `x = 1 +` is really an issue because the lexer emits a `Newline` (not a `NonLogicalNewline`) token and the parser correctly emits an error. The main challenge is compound statements that go over multiple cells, like this:

```py
if True:
```

```py
	print("true")
```

Emitting a `Newline` here isn't sufficient. But it might be enough to force emit a `Dedent`?

Overall, I feel like it should be enough if the lexer resets its lexing state (indent, open parentheses, interpolated string context) whenever it reaches a cell boundary, similar to what it already does when reaching the end of a file (which a cell boundary is). Things like multiline strings are a bit trickier but I don't think we have to solve all issues in a single pR

**Comment by spider-yamet** [comment]
Thanks for the clarification @MichaReiser  — this makes sense.
I’d like to take this issue and start with focused on the lexer-side cell-boundary reset you described, without trying to solve every notebook edge case at once.
At notebook cell boundaries (in Mode::Ipython), reset lexer state similarly to EOF handling:
indentation stack (flush with Dedent as needed),
open-parentheses / nesting state,
interpolated string context.
Keep the existing concatenated parse model (single AST), so lint/type-check behavior stays consistent.
Prioritize fixing cross-cell compound-statement continuation (if True: in one cell + indented block in next cell).

If this scope sounds right, I can start with this incremental implementation and follow up with additional edge cases in separate PRs.
Also,  I’d be happy to open a PR for this step. Thanks

**Comment by chinar-amrutkar** [comment]
@MichaReiser That makes sense. So the approach would be:

1. At each cell boundary, the lexer resets its state -forces a Dedent to indent level 0, closes any open parentheses in the token stream, and emits the appropriate Newline/Dedent sequence. This handles compound statements like if True: spanning cells because the forced Dedent closes the block.
2. Multiline strings and interpolation contexts would be flagged separately (or deferred as you mentioned).

I will start with the lexer state reset at cell boundaries and add test cases for the compound statement case. I will keep the scope of the first PR limited to this.

**Comment by MichaReiser** [comment]
@spider-yamet, your answer gives me some AI vibe. Please take a look at our [AI policy](https://github.com/astral-sh/.github/blob/main/AI_POLICY.md)

**Comment by omar-y-abdi** [comment]
@MichaReiser Is this issue still open and worth looking at or, has it been resolved already?

**Comment by MichaReiser** [comment]
We did some work in https://github.com/astral-sh/ruff/pull/24163 and explored different ideas, but the work stalled. I strongly suggest anyone to read through https://github.com/astral-sh/ruff/pull/24163 to consider the design discussion

**Comment by omar-y-abdi** [comment]
@MichaReiser 
Investigation summary on reproduction, PR test, and my approach:

Issue got reproduced on ruff 0.15.18. A compound statement that is unterminated inside its own cell gets silently completed by the next cell's indented body, so no error is reported.

I started with the tested approach in #24163 which reset lexer state at cell boundaries. 

The PR fixes the indentation case but I noticed it being incomplete in two ways:

1. It resets only indentation. Bracket nesting, triple-quoted strings, f-string state, and line continuations also carry across boundaries. Each is a separate masked error class, and all of them are still masked on 0.15.18.

2. It cannot catch grammar-level cross-cell errors at all. Example: a decorator in one cell with its `def` in the next. `@deco` NEWLINE `def` is valid Python, so no lexer boundary handling can produce that error.

I also explored an alternative: validate syntax per cell. Parse each cell standalone in the same IPython mode and collect ParseErrors only. A cell is valid if and only if it parses standalone, which matches how Jupyter executes cells.

I started stress testing my approach by treating each cell as a single-cell notebook, as an exact preview of the per-cell pass:

- No false positives on IPython syntax. Line magics, shell escapes, help, and top-level await all pass. Whole-cell magics are already excluded via `is_valid_python_code_cell`, so the per-cell pass inherits that exclusion.
- Catches every masked class, including the grammar-level decorator case the lexer reset provably cannot.
- Unicode safe, no panics. Cell offsets sit on valid UTF-8 boundaries.

Important caveat when implementing this: the per-cell pass must collect syntax errors only. Lint and semantic rules must stay on the concatenated source. 

Running them per cell would false-positive on every cross-cell reference. 

An example, `import os` alone reports F401, but `import os` followed by `print(os.getcwd())` is clean. 

Same problem for F821 on names defined in an earlier cell.

Additionally on performance: this adds one parse per cell, linear in total source, and can be reduced to a lex-only scan since it only needs ParseErrors. It shouldn’t regress the concatenated parse.

I validated this with the single-cell-notebook proxy on the current binary, not a built patch yet. If the direction looks right, I’m happy to implement the per-cell pass and open a PR.

**Comment by MichaReiser** [comment]
> Additionally on performance: this adds one parse per cell, linear in total source, and can be reduced to a lex-only scan since it only needs ParseErrors. It shouldn’t regress the concatenated parse.

Running parsing twice seems very expensive. Parsing is already one of the most expensive steps

**Comment by omar-y-abdi** [comment]
> > Additionally on performance: this adds one parse per cell, linear in total source, and can be reduced to a lex-only scan since it only needs ParseErrors. It shouldn’t regress the concatenated parse.
> 
> Running parsing twice seems very expensive. Parsing is already one of the most expensive steps

Two parses is too expensive, agreed. Better to make the single concatenated parse cell-aware instead, which is the direction #24163 was already going. The gap is that it only resets indentation, so the bracket, string, and f-string cases still leak across cells. Resetting the full lexer state at each boundary catches those too, all in one parse.

Ignore my lex-only suggestion. I was wrong there, the indent and bracket errors come from the parser, not the lexer.

Is that roughly what you had in mind for #24163?

**Comment by MichaReiser** [comment]
Yes, the idea was to cheaply handle it in the lexer. Whether that's doable, I don't know. I'm also not sure if there's any state in the parser that we'd need to reset at the cell boundary

**Comment by omar-y-abdi** [comment]
Checked the parser. Nothing persistent to reset at a boundary. `recovery_context` is saved and restored around every construct, and the depth guards are per-construct, so it's already clean between top-level statements.

So lexer-only should work for indent, bracket, string, f-string, and continuation. 
The one exception is decorators: `parse_decorators` eats `@deco` and the `def` as one unit, and a NEWLINE between them is legal, so the lexer can't split it. That one needs the parser to know the boundary. Rare, can wait.

**Comment by MichaReiser** [comment]
The parser's current stack contains some implicit state. E.g we need to make sure that it correctly exits from a `parse_binary_expression` before visting the next cell. But I think it should be sufficient to emit a `Newline` token at the end of the cell for that

**Comment by omar-y-abdi** [comment]
I agree, a logical Newline unwinds it out of parse_binary_expression. The catch is it has to be logical: nesting and string state still need resetting first, otherwise inside brackets or an f-string the lexer emits NL and won’t terminate.

**Comment by omar-y-abdi** [comment]
Should I put up a draft PR with the full lexer state reset plus the end-of-cell Newline, and leave the decorator case as a follow-up?

**Comment by MichaReiser** [comment]
Right... We could maybe fix this by emitting a cell boundary token in the lexer that we only accept in a few places.

Yeah, a PR is the easiest to have these kind of discussions

**Comment by MichaReiser** [comment]
The other alternative is to parse each cell separately, then concatenate the parsed modules together (still a single parse). The only thing we need to be careful about is to use `starts_at` to ensure the nodes use the correct offsets into the concatenated file. Maybe that's the easiest?

Which might be what you suggested. Sorry if I misunderstood you there

I think that's actually the easiest approach and comes with no overhead for regular files

<details>
<summary>Something along these lines</summary>

But with better names etc.

```rust
/// Parses each range as an independent module and combines the results into a single module.
///
/// This is useful for sources such as Jupyter notebooks, where every cell must be syntactically
/// valid on its own but definitions remain visible to subsequent cells. The ranges must be ordered
/// and non-overlapping.
pub fn parse_unchecked_module_ranges(
    source: &str,
    ranges: impl IntoIterator<Item = TextRange>,
    options: ParseOptions,
) -> Parsed<ModModule> {
    let mut ranges = ranges.into_iter().peekable();
    if ranges.peek().is_none() {
        return parse_unchecked(source, options)
            .try_into_module()
            .expect("module parse options must parse into a module");
    }

    let mut body = Suite::new();
    let mut tokens = Vec::new();
    let mut errors = Vec::new();
    let mut unsupported_syntax_errors = Vec::new();
    let mut module_range: Option<TextRange> = None;

    for range in ranges {
        if let Some(module_range) = module_range {
            debug_assert!(module_range.end() <= range.start());
        }

        let cell_source = &source[..range.end().to_usize()];
        let Parsed {
            syntax,
            tokens: cell_tokens,
            errors: cell_errors,
            unsupported_syntax_errors: cell_unsupported_syntax_errors,
        } = Parser::new_starts_at(cell_source, range.start(), options.clone())
            .parse()
            .try_into_module()
            .expect("module parse options must parse into a module");

        body.extend(syntax.body);
        tokens.extend(cell_tokens.iter().copied());
        errors.extend(cell_errors);
        unsupported_syntax_errors.extend(cell_unsupported_syntax_errors);
        module_range = Some(match module_range {
            Some(module_range) => TextRange::new(module_range.start(), range.end()),
            None => range,
        });
    }

    Parsed {
        syntax: ModModule {
            node_index: AtomicNodeIndex::NONE,
            range: module_range.expect("at least one range must be present"),
            body,
        },
        tokens: Tokens::new(tokens),
        errors,
        unsupported_syntax_errors,
    }
}

```
</details>

I wonder if some of the new synthesized newline tokens will get us into trouble with some downstream lint rules but we'll see

**Comment by MichaReiser** [comment]
@omar-y-abdi let me know if you're still interested in this, or if you'd be okay if I give the last suggestion a try

**Comment by omar-y-abdi** [comment]
@MichaReiser I've started working on the fix and will push a PR soon, sorry for the delay I had work during the week but i'm free now

**Comment by MichaReiser** [comment]
No worries. I just wanted to avoid that I start working on it when you already are. I'll assign the issue to you

**Comment by MichaReiser** [comment]
@omar-y-abdi you mentioned in the PR summary that it only fixes most of this issue. What are the remaining issues?

**Comment by MichaReiser** [comment]
I'll optimistically close this issue, as I'm not aware of cases that #26419 didn't fix.

**Comment by omar-y-abdi** [comment]
@MichaReiser You are correct in your thinking, no remaining syntax cases are left.

"Fixes part of" was conservative wording rather than a known gap. 
The one thing that intentionally stays on the concatenated source is cross-cell name resolution, a name defined in one cell and used in the next, which is correct Jupyter behavior. 

The only "part" not addressed in the PR is the deeper direction discussed earlier in the thread, making the single concatenated parse natively cell-aware instead of parsing each cell, which is an approach/performance refinement rather than an uncovered case.

You were right on closing this issue as complete.

On a side note, if you have any other issues were PR work has become stale, or where we can work together like we did on this PR, please feel free to assign the issue to me and i'll get to work.
